### PR TITLE
Remove undefined behavior in `value` method of boolean and primitive arrays

### LINKED ISSUE #645: Remove undefined behavior in `value` method of boolean and primitive arrays
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
As they OOB check uses `debug_assert` instead of `assert`, the check is not there in release mode.
We should use `assert!` instead.

**Describe the solution you'd like**
We should use `assert!` instead. In places where `value` is used directly and can be converted to use the unchecked variant, we can do so.

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by Dandandan**
# Which issue does this PR close?

Closes #645

# Rationale for this change

The value methods only had `debug_assert!` in them, which made that the check is removed in release mode.
 
 
 

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: arrow/src/array/array_boolean.rs [hunk 1]
```diff
@@ -100,7 +100,7 @@ impl BooleanArray {
     ///
     /// Note this doesn't do any bound checking, for performance reason.
     pub fn value(&self, i: usize) -> bool {
-        debug_assert!(i < self.len());
```
Comment (Dandandan) [inline arrow/src/array/array_boolean.rs:103]: UB 1

File: arrow/src/array/array_primitive.rs [hunk 2]
```diff
@@ -101,12 +101,10 @@ impl<T: ArrowPrimitiveType> PrimitiveArray<T> {
 
     /// Returns the primitive value at index `i`.
     ///
-    /// Note this doesn't do any bound checking, for performance reason.
-    /// # Safety
-    /// caller must ensure that the passed in offset is less than the array len()
+    /// Panics of offset `i` is out of bounds
     #[inline]
     pub fn value(&self, i: usize) -> T::Native {
-        debug_assert!(i < self.len());
```
Comment (Dandandan) [inline arrow/src/array/array_primitive.rs:109]: UB 2

File: arrow/src/array/array_string.rs [hunk 3]
```diff
@@ -103,28 +104,12 @@ impl<OffsetSize: StringOffsetSizeTrait> GenericStringArray<OffsetSize> {
     }
 
     /// Returns the element at index `i` as &str
+    #[inline]
     pub fn value(&self, i: usize) -> &str {
         assert!(i < self.data.len(), "StringArray out of bounds access");
-        //Soundness: length checked above, offset buffer length is 1 larger than logical array length
```
Comment (Dandandan) [inline arrow/src/array/array_string.rs:108]: Cleanup

File: arrow/src/compute/kernels/comparison.rs [hunk 4]
```diff
@@ -46,7 +46,10 @@ macro_rules! compare_op {
         let null_bit_buffer =
             combine_option_bitmap($left.data_ref(), $right.data_ref(), $left.len())?;
 
-        let comparison = (0..$left.len()).map(|i| $op($left.value(i), $right.value(i)));
+        // Safety:
+        // `i < $left.len()` and $left.len() == $right.len()
+        let comparison = (0..$left.len())
```
Comment (Dandandan) [inline arrow/src/compute/kernels/comparison.rs:51]: Avoid perf. penalty by using unsafe version here

File: arrow/src/compute/kernels/comparison.rs [hunk 5]
```diff
@@ -121,8 +124,10 @@ macro_rules! compare_op_primitive {
 macro_rules! compare_op_scalar {
     ($left: expr, $right:expr, $op:expr) => {{
         let null_bit_buffer = $left.data().null_buffer().cloned();
-
-        let comparison = (0..$left.len()).map(|i| $op($left.value(i), $right));
+        // Safety:
```
Comment (Dandandan) [inline arrow/src/compute/kernels/comparison.rs:127]: Avoid perf. penalty by using unsafe version here

**Review by paddyhoran — APPROVED** [review]
LGTM, thanks @Dandandan

**Review by alamb — APPROVED** [review]
Seems like a nice cleanup to me. @jorgecarleitao  would you like to review this PR as well?

**Review by jorgecarleitao — APPROVED** [review]
Yeap, good improvements. Thanks!

**Review by nevi-me — APPROVED** [review]

**Pull request merged**